contracts/GuardiansWallet.sol

```diff
@@ -0,0 +1,195 @@
+        uint split;
+    }
+    mapping (address => DistributorBatchState) distributorBatchState;
+
+    function isDelegatorRewardsBelowThreshold(uint256 delegatorRewards, uint256 totalRewards) private view returns (bool) {
+        return delegatorRewards.mul(100000) <= uint(maxDelegatorsStakingRewardsPercentMille).mul(totalRewards.add(toUint256Granularity(1))); // +1 is added to account for rounding errors
+    }
+
+    struct DistributeOrbsTokenStakingRewardsVars {
+        bool firstTxBySender;
+        address guardianAddr;
+        uint256 delegatorsAmount;
+    }
+    function distributeStakingRewards(uint256 totalAmount, uint256 fromBlock, uint256 toBlock, uint split, uint txIndex, address[] calldata to, uint256[] calldata amounts) external {
+        require(to.length > 0, "list must contain at least one recipient");
+        require(to.length == amounts.length, "expected to and amounts to be of same length");
+        uint48 totalAmount_uint48 = toUint48Granularity(totalAmount);
+        require(totalAmount == toUint256Granularity(totalAmount_uint48), "totalAmount must divide by 1e15");
+
+        DistributeOrbsTokenStakingRewardsVars memory vars;
+
+        vars.guardianAddr = getGuardiansRegistrationContract().resolveGuardianAddress(msg.sender);
+
+        for (uint i = 0; i < to.length; i++) {
+            if (to[i] != vars.guardianAddr) {
+                vars.delegatorsAmount = vars.delegatorsAmount.add(amounts[i]);
+            }
+        }
+        require(isDelegatorRewardsBelowThreshold(vars.delegatorsAmount, totalAmount), "Total delegators reward (to[1:n]) must be less then maxDelegatorsStakingRewardsPercentMille of total amount");
+
+        DistributorBatchState memory ds = distributorBatchState[vars.guardianAddr];
+        vars.firstTxBySender = ds.nextTxIndex == 0;
+
+        require(!vars.firstTxBySender || fromBlock == 0, "on the first batch fromBlock must be 0");
+
+        if (vars.firstTxBySender || fromBlock == ds.toBlock + 1) { // New distribution batch
+            require(txIndex == 0, "txIndex must be 0 for the first transaction of a new distribution batch");
+            require(toBlock < block.number, "toBlock must be in the past");
+            require(toBlock >= fromBlock, "toBlock must be at least fromBlock");
+            ds.fromBlock = fromBlock;
+            ds.toBlock = toBlock;
+            ds.split = split;
+            ds.nextTxIndex = 1;
+            distributorBatchState[vars.guardianAddr] = ds;
+        } else {
+            require(txIndex == ds.nextTxIndex, "txIndex mismatch");
+            require(toBlock == ds.toBlock, "toBlock mismatch");
+            require(fromBlock == ds.fromBlock, "fromBlock mismatch");
+            require(split == ds.split, "split mismatch");
+            distributorBatchState[vars.guardianAddr].nextTxIndex = txIndex + 1;
+        }
+
+        require(totalAmount_uint48 <= balances[vars.guardianAddr].stakingRewards, "not enough member balance for this distribution");
```

Comment:
Can read balances[vars.guardianAddr] to memory

Reply:
Right